app/org/sagebionetworks/bridge/services/AccountWorkflowService.java

```
@@ -0,0 +1,233 @@
+     * (neither is verified in this method).
+     */
+    public void sendEmailVerificationToken(Study study, String userId, String email) {
+        checkNotNull(study);
+        checkArgument(isNotBlank(userId));
+        checkArgument(isNotBlank(email));
+
+        String sptoken = createTimeLimitedToken();
+        
+        saveVerification(sptoken, new VerificationData(study.getIdentifier(), userId));
+        
+        String studyId = BridgeUtils.encodeURIComponent(study.getIdentifier());
+        String url = String.format(VERIFY_EMAIL_URL, BASE_URL, studyId, sptoken);
+        
+        BasicEmailProvider provider = new BasicEmailProvider.Builder()
+            .withStudy(study)
+            .withEmailTemplate(study.getVerifyEmailTemplate())
+            .withRecipientEmail(email)
+            .withToken(URL_TOKEN, url).build();
+        sendMailService.sendEmail(provider);         
+    }
+    
+    /**
+     * Send another email verification token. This creates and sends a new verification token 
+     * starting with the user's email address.
+     */
+    public void resendEmailVerificationToken(StudyIdentifier studyIdentifier, Email email) {
+        checkNotNull(studyIdentifier);
+        checkNotNull(email);
+        
+        Study study = studyService.getStudy(studyIdentifier);
+        Account account = accountDao.getAccountWithEmail(study, email.getEmail());
+        if (account != null) {
+            sendEmailVerificationToken(study, account.getId(), email.getEmail());
+        }
+    }
+    
+    /**
+     * Using the verification token that was sent to the user, verify the email address. 
+     * If the token is invalid, it fails quietly. If the token exists but the account 
+     * does not, it throws an exception (this would be unexpected).
+     */
+    public void verifyEmail(EmailVerification verification) {
+        checkNotNull(verification);
+
+        VerificationData data = restoreVerification(verification.getSptoken());
```

Could we have it fail on a non-existent SpToken? My thought is that failing silently could make debugging for devs and reporting errors for customers frustrating since there is no sign at the point of failure.

Guessing the correct token should be on the same order of difficulty as guessing the correct email plus password, and I don't think getting an error leaks any info since the token itself shouldn't have any semantic meaning that could identify the user, the probability of getting this error is approximately 100% without a real token.

---

This is not the current behavior but I agree with you that it's best to return a 400 here. Changed it.